design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,421 @@
+schema information. As hosts of the same vendor and model will use
+the same schema, its not necessary to cache it each time for similar hardware.
+When BIOS data is read from the Ironic API for each host the following
+actions will be taken:
+
+- If the `HostFirmwareSettings` already has a reference to a  `FirmwareSchema`
+  resource, then the schema data in the BIOS API will be ignored.
+- If there is no reference, the schema data for all BIOS settings for the
+  host will be stored locally.
+- A hash will be generated from this stored data, if a corresponding hash
+  does not exist in the `FirmwareSchema` CRs, a new CR will be created and
+  the schema data will be copied to it.
+- The reference to the `FirmwareSchema` resource in `HostFirmwareSettings`
+  will be updated with the hash. A suggested reference string is to
+  include the vendor and model in the string, such as
+  "vendor,model,64 chars of hash". The vendor and model can be obtained
+  from the associated BareMetalHost resource.
+- Using the hash will ensure that the schemas match. In the case of
+  hardware from the same vendor and model, but with a different firmware
+  version, it is possible that the schemas would differ. In this case a new
+  `FirmwareSchema` will be created.
+- The name of the `FirmwareSchema` resource will be added to the Labels
+  for the `HostFirmwareSettings` so that it can be used to delete the
+  `FirmwareSchema` resource when it is no longer being used.
+- It is possible that the hosts will be in different namespaces (and the
+  `FirmwareSchema` will be in same namespace as host). For this case,
+  a `namespace` hint has been added to the `HostFirmwareSettings` CRD
+  to specify the namespace to use for the schema.
+
+An example of the proposed CRD for `HostFirmwareSettings` is as follows:
+
+```yaml
+---
+kind: CustomResourceDefinition
+metadata:
+  name: hostfirmwaresettings.metal3.io
+spec:
+  group: metal3.io
+  names:
+    kind: HostFirmwareSettings
+    listKind: HostFirmwareSettingsList
+  scope: Namespaced
+  versions:
+  - name: v1alpha1
+    schema:
+          spec:
+            description: HostFirmwareSettingsSpec defines the desired state
+            properties:
+              vendorSettings:
+                items:
+                  description: describes a setting whose value can be changed
+                  properties:
+                    name:
+                      description: Identifier for the Firmware setting.
+                      type: string
+                    value:
+                      description: Value of the Firmware setting.
+                      type: string
+                  required:
+                  - name
+                  - value
+                  type: object
+                type: array
+            type: object
+          status:
+            description: defines the observed state of HostFirmwareSettings
+            properties:
+              schemaReference:
```

Comment:
https://github.com/kubernetes-sigs/cluster-api/issues/2318 has some useful tips on object reference types.

Reply:
Thanks Doug. IIUC from that we should define a separate struct for the reference such as in https://github.com/kubernetes/api/blob/release-1.17/admissionregistration/v1/types.go#L533, and not use corev1.ObjectReference.  I'll make a note of that in the doc.